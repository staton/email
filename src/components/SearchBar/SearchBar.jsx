import React, { Component } from 'react';
import SearchBarButton from './SearchBarButton/SearchBarButton';
import SearchBarInput from './SearchBarInput/SearchBarInput';
import STRINGS from '../../resources/strings';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { MdClear, MdSearch  } from 'react-icons/md';
import { setSearchText } from '../../redux/actions/app-actions';

export class SearchBar extends Component {

    constructor() {
        super();
        
        // In this component, local state will be used to determine the 
        // style of the search bar, based on whether or not it is in focus.
        this.state = { isSearchFocused: false };

        this.handleClearSearchText = this.handleClearSearchText.bind(this);
        this.handleSearch = this.handleSearch.bind(this);
        this.handleSearchBlurred = this.handleSearchBlurred.bind(this);
        this.handleSearchFocused = this.handleSearchFocused.bind(this);
        this.handleSearchKeyDown = this.handleSearchKeyDown.bind(this);
        this.handleSearchTextChanged = this.handleSearchTextChanged.bind(this);
    }

	render() {
        const innerDivClassNames = this.getSearchBarInnerDivClassNames();

		return (
            <div 
                className="SearchBar"
                style={{ backgroundColor: this._backgroundColor }}
            >
                <div className={innerDivClassNames}>
                    <SearchBarButton 
                        content={<MdSearch />}
                        onClick={this.search}
                        title={STRINGS.Search}
                    />
                    <SearchBarInput 
                        onBlur={this.handleSearchBlurred}
                        onChange={this.handleSearchTextChanged}
                        onFocus={this.handleSearchFocused}
                        onKeyDown={this.handleSearchKeyDown}
                    />
                    {
                        // only display the clear-search button if the search bar contains text
                        (this.props.searchText)
                        ?   <SearchBarButton
                                content={<MdClear />}
                                onClick={this.handleClearSearchText}
                                title={STRINGS.Clear}
                            />
                        :   null
                    }
                </div>
            </div>
		);
    }
    
    /**
     * Gets the class names for the search bar, based on if the input is focused or not.
     * @returns {string} The class names for the search bar.
     */
    getSearchBarInnerDivClassNames() {
        const searchBarInnerDivClassName = 'SearchBar__inner-div';
        const searchBarInnerDivFocusedClassNames = searchBarInnerDivClassName + ' SearchBar__inner-div--focused';

        return (this.state.isSearchFocused)
            ? searchBarInnerDivFocusedClassNames
            : searchBarInnerDivClassName;
    }
    
    /**
     * Clears the current search bar input text.
     * @param {object} e The event object.
     */
    handleClearSearchText(e) {
        if (this.props.searchText) {
            this.props.setSearchText('');
        }
    }

    /**
     * Performs a search based on the current search bar input text.
     * @param {object} e The event object.
     */
    handleSearch(e) {
        if (this.props.searchText) {

        }
    }

    /**
     * Called when the search input loses focus.
     * @param {object} e The event object.
     */
    handleSearchBlurred(e) {
        // remove the focused class name from this element, to revert to the
        // search bar's default style.
        this.setState({ isSearchFocused: false });
    }

    /**
     * Called when the search input receives focus.
     * @param {object} e The event object.
     */
    handleSearchFocused(e) {
        // add the focused class name to this element, which will control
        // the search bar style when it is focused.
        this.setState({ isSearchFocused: true });
    }

    /**
     * Called when there is a key-down event in the search input.
     * @param {object} e The event object.
     */
    handleSearchKeyDown(e) {
        
    }

    /**
     * Called when the search input changes.
     * @param {object} e The event object.
     */
    handleSearchTextChanged(e) {
        if (e.which === 13) {
            console.log('enter pressed!!!!!!');
        } else if (this.props.searchText !== e.target.value) {
            this.props.setSearchText(e.target.value);
        }
    }

}

function mapStateToProps(store, ownProps) {
    return {
        searchText: store.app.searchText
    };
}

function mapDispatchToProps(dispatch) {
    return bindActionCreators({
        setSearchText: setSearchText
    },
    dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(SearchBar);