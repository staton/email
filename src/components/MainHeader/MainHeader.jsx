import React, { Component } from 'react';
import PropTypes from 'prop-types';
import HeaderButton from '../HeaderButton/HeaderButton';
import SearchBar from '../SearchBar/SearchBar';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { MdMenu, MdSearch, MdSettings  } from 'react-icons/md';
import { setDrawerVisibility } from '../../redux/actions/app-actions';
import { setSearchBarVisibility } from '../../redux/actions/search-actions';

const propTypes = {
    leftElementGroupClassName: PropTypes.string.isRequired,
    rightElementGroupClassName: PropTypes.string.isRequired,
    elementClassName: PropTypes.string.isRequired
};

const defaultProps = {
};

export class MainHeader extends Component {

    constructor() {
        super();

        this.handleDrawerButtonClicked = this.handleDrawerButtonClicked.bind(this);
        this.handleSearchButtonClicked = this.handleSearchButtonClicked.bind(this);
        this.handleSettingsButtonClicked = this.handleSettingsButtonClicked.bind(this);
    }

	render() {
        const logoClassName = 'MainHeader__logo';
        const menuButtonId = 'MainHeader__menu-button';
        const searchButtonId = 'MainHeader__search-button';
        const settingsButtonId = 'MainHeader__settings-button';

		return (
			<header className="MainHeader">
                <div className={this.props.leftElementGroupClassName}>
                    <div className={this.props.elementClassName}>
                        <HeaderButton
                            content={<MdMenu />}
                            id={menuButtonId}
                            onClick={this.handleDrawerButtonClicked} 
                        />
                    </div>
                    <div className={this.props.elementClassName + ' ' + logoClassName}>
                        logo
                    </div>
                    {(this.props.isSmallScreen) ? null : this.getSearchBar()}
                </div>
                <div className={this.props.rightElementGroupClassName}>
                    {(this.props.isSmallScreen) ? this.getSearchButton(searchButtonId) : null}
                    <div className={this.props.elementClassName}>
                        <HeaderButton
                            content={<MdSettings />}
                            id={settingsButtonId}
                            onClick={this.handleSettingsButtonClicked} 
                        />
                    </div>
                </div>
			</header>
		);
    }
    
    /**
     * Gets the search bar element and its wrapper.
     * @returns {Element} The search bar element.
     */
    getSearchBar() {
        return (
            <div className={this.props.elementClassName}>
                <SearchBar />
            </div>
        );
    }

    /**
     * Gets the search button element and its wrapper.
     * @param {string} id The id of the search button.
     * @returns {Element} The search button element.
     */
    getSearchButton(id) {
        return (
            <div className={this.props.elementClassName}>
                <HeaderButton
                    content={<MdSearch />}
                    id={id}
                    onClick={this.handleSearchButtonClicked} 
                />
            </div>
        );
    }

    /**
     * Called when the user clicks the open/close drawer button.
     * @param {object} e The event.
     */
    handleDrawerButtonClicked(e) {
        this.props.setDrawerVisibility(!this.props.isDrawerVisible);
    }

    /**
     * Called when the user clicks the search button in the header.
     * @param {object} e The event.
     */
    handleSearchButtonClicked(e) {
        this.props.setSearchBarVisibility(!this.props.isSearchBarVisible);
    }

    /**
     * Called when the user clicks on the settings button.
     * @param {object} e The event.
     */
    handleSettingsButtonClicked(e) {
        
    }

}

MainHeader.propTypes = propTypes;
MainHeader.defaultProps = defaultProps;

function mapStateToProps(store, ownProps) {
    return {
        isDrawerVisible: store.app.isDrawerVisible,
        isSearchBarVisible: store.search.isSearchBarVisible,
        isSmallScreen: store.app.isSmallScreen
    };
}

function mapDispatchToProps(dispatch) {
    return bindActionCreators({
        setDrawerVisibility: setDrawerVisibility,
        setSearchBarVisibility: setSearchBarVisibility
    },
    dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(MainHeader);