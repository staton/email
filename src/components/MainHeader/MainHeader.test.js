import React from 'react';
import toJson from 'enzyme-to-json';
import SearchBar from '../SearchBar/SearchBar';
import { shallow } from 'enzyme';
import { MainHeader } from './MainHeader';

describe('<MainHeader />', () => {

    it('renders correctly for small screens', () => {
        const wrapper = shallow(
            <MainHeader 
                leftElementGroupClassName="Header__left-element-group"
                rightElementGroupClassName="Header__right-element-group"
                elementClassName="Header__element"
                isSmallScreen={true}
            />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders correctly for large screens', () => {
        const wrapper = shallow(
            <MainHeader 
                leftElementGroupClassName="Header__left-element-group"
                rightElementGroupClassName="Header__right-element-group"
                elementClassName="Header__element"
                isSmallScreen={false} 
            />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    
    it('renders menu btn, search btn, settings btn, logo, and search bar for large screens', () => {
        const wrapper = shallow(
            <MainHeader 
                leftElementGroupClassName="Header__left-element-group"
                rightElementGroupClassName="Header__right-element-group"
                elementClassName="Header__element"
                isSmallScreen={false} 
            />);
        expect(wrapper.find('#MainHeader__menu-button').length).toEqual(1);
        expect(wrapper.find('#MainHeader__search-button').length).toEqual(0);
        expect(wrapper.find('#MainHeader__settings-button').length).toEqual(1);
        expect(wrapper.find('.MainHeader__logo').length).toEqual(1);
        expect(wrapper.find(SearchBar).length).toEqual(1);
    });

    it('correctly calls setDrawerVisibility when the user clicks on the menu button when drawer is closed (small screens)', () => {
        const isDrawerVisible = false;  // drawer is closed
        const setDrawerVisibility = jest.fn();
        const wrapper = shallow(
            <MainHeader 
                leftElementGroupClassName="Header__left-element-group"
                rightElementGroupClassName="Header__right-element-group"
                elementClassName="Header__element"
                isDrawerVisible={isDrawerVisible}
                isSmallScreen={true}
                setDrawerVisibility={setDrawerVisibility} 
            />);
        wrapper.find('#MainHeader__menu-button').simulate('click');
        expect(setDrawerVisibility).toHaveBeenCalledTimes(1);
        expect(setDrawerVisibility).toHaveBeenCalledWith(!isDrawerVisible);
    });

    it('correctly calls setDrawerVisibility when the user clicks on the menu button when drawer is closed (large screens)', () => {
        const isDrawerVisible = false;  // drawer is closed
        const setDrawerVisibility = jest.fn();
        const wrapper = shallow(
            <MainHeader 
                leftElementGroupClassName="Header__left-element-group"
                rightElementGroupClassName="Header__right-element-group"
                elementClassName="Header__element"
                isDrawerVisible={isDrawerVisible}
                isSmallScreen={false}
                setDrawerVisibility={setDrawerVisibility} 
            />);
        wrapper.find('#MainHeader__menu-button').simulate('click');
        expect(setDrawerVisibility).toHaveBeenCalledTimes(1);
        expect(setDrawerVisibility).toHaveBeenCalledWith(!isDrawerVisible);
    });

    it('correctly calls setDrawerVisibility when the user clicks on the menu button when drawer is open (large screens)', () => {
        const isDrawerVisible = true;   // drawer is open
        const setDrawerVisibility = jest.fn();
        const wrapper = shallow(
            <MainHeader 
                leftElementGroupClassName="Header__left-element-group"
                rightElementGroupClassName="Header__right-element-group"
                elementClassName="Header__element"
                isDrawerVisible={isDrawerVisible}
                isSmallScreen={false}
                setDrawerVisibility={setDrawerVisibility} 
            />);
        wrapper.find('#MainHeader__menu-button').simulate('click');
        expect(setDrawerVisibility).toHaveBeenCalledTimes(1);
        expect(setDrawerVisibility).toHaveBeenCalledWith(!isDrawerVisible);
    });

    it('correctly calls setSearchBarVisibility when the user clicks on the search button when search bar is hidden', () => {
        const isSearchBarVisible = false;   // search bar is hidden
        const setSearchBarVisibility = jest.fn();
        const wrapper = shallow(
            <MainHeader 
                leftElementGroupClassName="Header__left-element-group"
                rightElementGroupClassName="Header__right-element-group"
                elementClassName="Header__element"
                isSearchBarVisible={isSearchBarVisible}
                isSmallScreen={true}
                setSearchBarVisibility={setSearchBarVisibility} 
            />);
        wrapper.find('#MainHeader__search-button').simulate('click');
        expect(setSearchBarVisibility).toHaveBeenCalledTimes(1);
        expect(setSearchBarVisibility).toHaveBeenCalledWith(!isSearchBarVisible);
    });

    it('correctly calls setSearchBarVisibility when the user clicks on the search button when search bar is open', () => {
        const isSearchBarVisible = true;   // search bar is open
        const setSearchBarVisibility = jest.fn();
        const wrapper = shallow(
            <MainHeader 
                leftElementGroupClassName="Header__left-element-group"
                rightElementGroupClassName="Header__right-element-group"
                elementClassName="Header__element"
                isSearchBarVisible={isSearchBarVisible}
                isSmallScreen={true}
                setSearchBarVisibility={setSearchBarVisibility} 
            />);
        wrapper.find('#MainHeader__search-button').simulate('click');
        expect(setSearchBarVisibility).toHaveBeenCalledTimes(1);
        expect(setSearchBarVisibility).toHaveBeenCalledWith(!isSearchBarVisible);
    });

});