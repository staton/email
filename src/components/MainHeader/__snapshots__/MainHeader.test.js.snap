// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainHeader /> renders correctly for large screens 1`] = `
<header
  className="MainHeader"
>
  <div
    className="Header__left-element-group"
  >
    <div
      className="Header__element"
    >
      <HeaderButton
        content={<Unknown />}
        enabled={true}
        id="MainHeader__menu-button"
        onClick={[Function]}
      />
    </div>
    <div
      className="Header__element MainHeader__logo"
    >
      logo
    </div>
    <div
      className="Header__element"
    >
      <Connect(SearchBar) />
    </div>
  </div>
  <div
    className="Header__right-element-group"
  >
    <div
      className="Header__element"
    >
      <HeaderButton
        content={<Unknown />}
        enabled={true}
        id="MainHeader__settings-button"
        onClick={[Function]}
      />
    </div>
  </div>
</header>
`;

exports[`<MainHeader /> renders correctly for small screens 1`] = `
<header
  className="MainHeader"
>
  <div
    className="Header__left-element-group"
  >
    <div
      className="Header__element"
    >
      <HeaderButton
        content={<Unknown />}
        enabled={true}
        id="MainHeader__menu-button"
        onClick={[Function]}
      />
    </div>
    <div
      className="Header__element MainHeader__logo"
    >
      logo
    </div>
  </div>
  <div
    className="Header__right-element-group"
  >
    <div
      className="Header__element"
    >
      <HeaderButton
        content={<Unknown />}
        enabled={true}
        id="MainHeader__search-button"
        onClick={[Function]}
      />
    </div>
    <div
      className="Header__element"
    >
      <HeaderButton
        content={<Unknown />}
        enabled={true}
        id="MainHeader__settings-button"
        onClick={[Function]}
      />
    </div>
  </div>
</header>
`;
